src/merge_stop_areas.rs

```diff
@@ -0,0 +1,338 @@
+//
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see
+// <http://www.gnu.org/licenses/>.
+
+//! [NTFS](https://github.com/CanalTP/navitia/blob/dev/documentation/ntfs/ntfs_fr.md)
+//! format management.
+
+extern crate serde;
+extern crate serde_json;
+
+use collection::CollectionWithId;
+use csv;
+use failure::ResultExt;
+use model::Collections;
+use objects::StopArea;
+use objects::{CommentLinksT, KeysValues};
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::fs;
+use std::path;
+use std::path::PathBuf;
+use std::result::Result as StdResult;
+use Result;
+
+#[derive(Deserialize, Debug)]
+struct StopAreaMergeRule {
+    #[serde(rename = "stop_area_id")]
+    id: String,
+    #[serde(rename = "stop_area_name")]
+    name: String,
+    group: String,
+    priority: u16,
+}
+
+#[derive(Debug, Clone, Eq)]
+struct StopAreaGroupRule {
+    master_stop_area_id: String,
+    to_merge_stop_area_ids: Vec<String>,
+}
+
+impl StopAreaGroupRule {
+    fn ensure_rule_valid(
+        self,
+        stop_area_ids: &[String],
+        report: &mut MergeStopAreasReport,
+    ) -> Result<Self> {
+        let mut valid_rule = self.clone();
+        let message = format!(
+            "rule for master {} contains unexisting stop areas and is no longer valid",
+            self.master_stop_area_id
+        );
+        valid_rule
+            .to_merge_stop_area_ids
+            .retain(|id| stop_area_ids.contains(&id));
+        if valid_rule.to_merge_stop_area_ids.is_empty() {
+            report.add_error(
+                format!(
+                    "rule for master {} does not contain any existing stop areas to merge",
+                    self.master_stop_area_id
+                ),
+                MergeStopAreasReportType::NothingToMerge,
+            );
+            bail!(message);
+        } else if !stop_area_ids.contains(&valid_rule.master_stop_area_id) {
+            if valid_rule.to_merge_stop_area_ids.len() == 1 {
+                report.add_error(
+                    format!("master {} of rule does not exist anymore and cannot be replaced by an other one", self.master_stop_area_id),
+                     MergeStopAreasReportType::NoMasterPossible);
+                bail!(message);
+            }
+            report.add_warning(
+                format!(
+                    "master {} of rule does not exist and has been replaced by an other one",
+                    self.master_stop_area_id
+                ),
+                MergeStopAreasReportType::MasterReplaced,
+            );
+            valid_rule.master_stop_area_id = valid_rule.to_merge_stop_area_ids.remove(0);
```

Comment:
hum, it was not in specs :D
@prhod should I add a warning?

Reply:
I think your implementation may remove the master in this case.

Reply:
I didn't think of a rule merge with the master id included in the children ids ... It could be a problem.
Can you add a check and log it as an error ?

Reply:
In fact I don't think it's possible by design (see function group_rules_from_file_rules) the master is selected from the stop areas by priority and cannot be in the stop areas to merge

Reply:
OK